lib/Sema/TypeCheckInvertible.cpp

```diff
@@ -349,16 +349,26 @@ ProtocolConformance *deriveConformanceForInvertible(Evaluator &evaluator,
   // The conformanceDC specifies THE decl context to use for the conformance.
   auto generateConformance =
       [&](DeclContext *conformanceDC) -> ProtocolConformance * {
-    // Form a conformance.
-    auto conformance = ctx.getNormalConformance(
-        nominal->getDeclaredInterfaceType(), proto, nominal->getLoc(),
-        conformanceDC, ProtocolConformanceState::Complete,
-        /*isUnchecked=*/false, /*isPreconcurrency=*/false);
-    conformance->setSourceKindAndImplyingConformance(
-        ConformanceEntryKind::Synthesized, nullptr);
-
-    nominal->registerProtocolConformance(conformance, /*synthesized=*/true);
-    return conformance;
+    if (!isa<ClassDecl>(nominal)) {
+      // Form a conformance.
+      auto conformance = ctx.getNormalConformance(
+          nominal->getDeclaredInterfaceType(), proto, nominal->getLoc(),
+          conformanceDC, ProtocolConformanceState::Complete,
+          /*isUnchecked=*/false, /*isPreconcurrency=*/false);
+      conformance->setSourceKindAndImplyingConformance(
+          ConformanceEntryKind::Synthesized, nullptr);
+
+      nominal->registerProtocolConformance(conformance, /*synthesized=*/true);
+      return conformance;
+    } else {
+      // Classes need to return a built-in conformance to avoid cycles with
+      // superclasses, such as:
+      //    class Base<T>: IP {}
+      //    class Derived: Base<Derived> {}
+      return ctx.getBuiltinConformance(nominal->getDeclaredInterfaceType(),
```

Comment:
the conformance should be formed in lookupConformance() with the conforming type, not the nominal's declared interface type. Generally we want this to hold:

```
lookupConformance(type, proto)->getType()->isEqual(type)
```